File: python/218. The Skyline Problem.py
class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        tmp = []

        for i in range(len(buildings)):
            start, end, height = buildings[i]
            tmp += [(start, height, i, True)]
            tmp += [(end, height, i, False)]

        maxheap = MaxHeap()
        intervals = []
        lastpos = None
        for pos, height, index, isStart in sorted(tmp):
            maxHeight = maxheap.top[0] if len(maxheap) else 0
            self.mergeTo(intervals, lastpos, pos, maxHeight)
            if isStart:
                maxheap.push((height, index))
            else:
                maxheap.pop((height, index))
            lastpos = pos

        res = []
        for start, end, height in intervals:
            if res and res[-1][0] == start:
                res.pop()
            res += [[start, height]]
            res += [[end, 0]]
        # print(intervals)
        return res

    # lint code
    def mergeTo(self, intervals, start, end, height):
        if start is None or height == 0 or start == end:
            return
        if not intervals:
            intervals.append([start, end, height])
            return
        _, prevEnd, prevHeight = intervals[-1]
        if prevHeight == height and prevEnd == start:
            intervals[-1][1] = end
            return
        intervals.append([start, end, height])


class HashHeap:
    def __init__(self):
        self.heap = []
        self.map = {}

    def __len__(self):
        return len(self.heap)

    @property
    def top(self):
        return self.heap[0]

    @property
    def last(self):
        return len(self) - 1

    def push(self, val):
        self.heap += [val]
        self.map[val] = self.last
        self._siftup(self.last)

    def pop(self, val=None):
        if not val:
            val = self.top
        if val not in self.map:
            return

        # move to last
        idx = self.map[val]
        self._swap(idx, self.last)

        # then pop
        del self.map[val]
        self.heap.pop()

        if idx != len(self):
            self._siftup(idx)
            self._siftdown(idx)

        return val

    def _swap(self, i, j):
        tmp1, tmp2 = self.heap[i], self.heap[j]
        self.heap[i], self.heap[j] = tmp2, tmp1
        self.map[tmp1], self.map[tmp2] = j, i

    def _siftup(self, idx):
        while idx != 0:
            parent = (idx - 1) // 2
            if parent == self._findParent([parent, idx]):
                return
            self._swap(parent, idx)
            idx = parent

    def _siftdown(self, idx):
        while idx * 2 < len(self):
            left, right = idx * 2 + 1, idx * 2 + 2
            parent = self._findParent([left, right, idx])
            if idx == parent:
                return
            self._swap(parent, idx)
            idx = parent


class MinHeap(HashHeap):
    def _findParent(self, canadiates):
        return min(
            canadiates, key=lambda i: self.heap[i][0] if i < len(self) else sys.maxsize
        )


class MaxHeap(HashHeap):
    def _findParent(self, canadiates):
        return max(
            canadiates, key=lambda i: self.heap[i][0] if i < len(self) else -sys.maxsize
        )

File: swift/218. The Skyline Problem.swift
class Solution {
    func getSkyline(_ buildings: [[Int]]) -> [[Int]] {
        var tmp:[(pos:Int, height:Int, idx: Int, isStart: Bool)] = []
        for i in 0..<buildings.count {
            let start = buildings[i][0]
            let end = buildings[i][1]
            let height = buildings[i][2]
            
            tmp += [(start, height, i, true)]
            tmp += [(end, height, i, false)]
        }
        
        var maxHeap = HashHeap<Int>(false)
        var intervals:[(start: Int, end: Int, height: Int)] = []
        var lastPos: Int?
        
        func merge(_ start: Int?, _ end: Int, _ height: Int) {
            guard let start = start, start != end && height != 0 else { return }
            
            if let last = intervals.last, last.end == start && last.height == height {
                intervals[intervals.count - 1] = (last.start, end, height)
                return
            }
            
            intervals += [(start, end, height)]
        }
        
        for (pos, height, idx, isStart) in tmp.sorted { ($0.pos, $0.height, $0.idx) < ($1.pos, $1.height, $1.idx) } {
            merge(lastPos, pos, maxHeap.top ?? 0)
            lastPos = pos
            
            isStart ? maxHeap.push(height, idx) : maxHeap.pop(height, idx)
        }
        
        var res: [[Int]] = []
        
        for (start, end, height) in intervals {
            if let last = res.last, last[0] == start && last[1] == 0 {
                res.removeLast()
            }
            
            res += [[start, height]]
            res += [[end, 0]]
        }
        
        return res
    }
}

class HashHeap<T: Hashable & Comparable> {
    private struct Node: Hashable {
        var val: T
        var identifier: Int
        init(_ val: T, _ identifier: Int = 0) {
            self.val = val
            self.identifier = identifier
        }
    }
    private var heap: [Node] = []
    private var map: [Node: Int] = [:]
    private var size: Int { heap.count }
    private let isMinHeap: Bool
    var top: T? { heap.first?.val }
    
    init(_ isMinHeap: Bool) {
        self.isMinHeap = isMinHeap
    }
    
    func push(_ val: T, _ identifier: Int = 0) { push(Node(val, identifier)) }

    func pop(_ val: T, _ identifier: Int = 0) { return pop((Node(val, identifier)))}
    
    func pop() -> T? {
        guard let top = heap.first else { return nil }
        pop(top) 
        return top.val
    }
    
    private func push(_ node: Node) {
        heap += [node]
        map[node] = size - 1
        siftUp(size - 1)
    }
    
    private func pop(_ node: Node) {
        guard let idx = map[node] else { return }
        swap(idx, size - 1)
        heap.removeLast()
        map[node] = nil
        
        if idx < size {
            siftUp(idx)
            siftDown(idx)
        }
    }
    
    private func siftUp(_ idx: Int) {
        guard idx > 0 else { return }
        let parent = (idx - 1) / 2
        guard parent != findParent([parent, idx]) else { return }
        swap(idx, parent)
        siftUp(parent)
    }
    
    private func siftDown(_ idx: Int) {
        guard idx * 2 < size else { return }
        guard 
            let parent = findParent([idx, idx * 2 + 1, idx * 2 + 2]),
            idx != parent else { return }
        swap(idx, parent)
        siftDown(parent)
    }
    
    private func swap(_ i: Int, _ j: Int) {
        (map[heap[i]], map[heap[j]]) = (j, i)
        heap.swapAt(i, j)
    }
    
    private func findParent(_ nums: [Int]) -> Int? {
        let sorted = nums.filter { $0 < heap.count }.sorted { heap[$0].val < heap[$1].val }
        return isMinHeap ? sorted.first : sorted.last
    }
}